### PR TITLE
Add piecewise linear codec, deprecate linear and mulitinear,  complete benchmark with real world datasets.

### LINKED ISSUE #1215: Slope value unexpected in multilinear interpolation
```rust
  #[test]
  fn test_compression_on_simple_data() {
      let data = (0..=CHUNK_SIZE*11 as u64).collect::<Vec<_>>();
      let (estimate, actual_compression) =
          create_and_validate(&data, "simple monotonically");
      println!("actual compression {} estimate {} data len {}", actual_compression, estimate, data.len());
      assert!(actual_compression < 0.2);
      assert!(estimate < 0.20);
      assert!(estimate > 0.15);
      assert!(actual_compression > 0.1);
  }
```

I printed the slopes and the length of the interpolation

```
slope 1.0019569
slope 1.0019569
slope 1.0019569
slope 1.0019569
slope 1.0019569
slope 1.0019569
slope 1.0019569
slope 1.0019569
slope 1.0019569
slope 1.0019569
slope 1.0019569
slope NaN
interpolations len 12
```

I would have expected a slope of 1 and only 11 interpolation functions.

### PR DESCRIPTION
**Pull request by fmassot**
Ok PR is ready for review.

Here is a summary of the PR:
- Piecewise linear codec added: basically it's a refactor and it fixes #1215
- #1212 partially fixed that by saving 8 bytes.
- Linear interpolation and multilinear deprecated. It's now only there for the reading part so that users can still read these fast field codecs
- FOR codec added with feature `unstable`, it's not used in tantivy for the moment.
- Bench also compression time/reading time and on real-world datasets.

### CODE REVIEW

File: fastfield_codecs/src/multilinearinterpol_2.rs [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+use tantivy_bitpacker::BitPacker;
+
+use common::BinarySerializable;
+use common::DeserializeFrom;
+use tantivy_bitpacker::BitUnpacker;
+
+const BLOCK_SIZE: u64 = 512;
+
+#[derive(Clone)]
+pub struct MultiLinearInterpol2FastFieldReader {
```
Comment (PSeitz) [inline fastfield_codecs/src/multilinearinterpol_2.rs]:
```suggestion
pub struct MultiLinearInterpolV2FastFieldReader {
```

File: fastfield_codecs/README.md [hunk 2]
```diff
@@ -29,40 +40,108 @@ cargo run --features bin
...
++----------------------------------+-------------------+------------------------------+--------------------------+----------------------+
+| Monotonically increasing concave |                   |                              |                          |                      |
++----------------------------------+-------------------+------------------------------+--------------------------+----------------------+
+| LinearInterpol                   | 0.25003937        | 0.26562938                   | 775                      | 113                  |
++----------------------------------+-------------------+------------------------------+--------------------------+----------------------+
+| MultiLinearInterpol              | 0.05403           | 0.1883836                    | 914                      | 223                  |
++----------------------------------+-------------------+------------------------------+--------------------------+----------------------+
+| MultiLinearInterpol2             | 0.005955          | 0.18813984                   | 899                      | 245                  |
```
Comment (PSeitz) [inline fastfield_codecs/README.md]:
```suggestion
| MultiLinearInterpolV2             | 0.005955          | 0.18813984                   | 899                      | 245                  |
```

File: fastfield_codecs/src/multilinearinterpol_2.rs [hunk 3]
```diff
@@ -0,0 +1,353 @@
+/*!
+
+MultiLinearInterpol compressor uses linear interpolation to guess a values and stores the offset, but in blocks of 512.
+
+With a CHUNK_SIZE of 512 and 29 byte metadata per block, we get a overhead for metadata of 232 / 512 = 0,45 bits per element.
```
Comment (PSeitz) [inline fastfield_codecs/src/multilinearinterpol_2.rs]: Outdated comment. I think outlining the V2 changes would make sense

File: fastfield_codecs/src/multilinearinterpol_2.rs [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+    min_value: u64,
+    max_value: u64,
+    block_metadatas: Vec<BlockMetadata>,
+    block_start_offsets: Vec<u64>,
+    block_bit_unpackers: Vec<BitUnpacker>,
+}
+
+#[derive(Clone, Debug, Default)]
+struct BlockMetadata {
+    constant: u64,
```
Comment (PSeitz) [inline fastfield_codecs/src/multilinearinterpol_2.rs]: I don't think constant is a good name

File: fastfield_codecs/src/multilinearinterpol_2.rs [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+    num_bits: u8,
+}
+
+
+impl BinarySerializable for BlockMetadata {
+    fn serialize<W: Write>(&self, write: &mut W) -> io::Result<()> {
+        self.constant.serialize(write)?;
+        self.constant_positive_offset.serialize(write)?;
+        self.slope.serialize(write)?;
+        self.num_bits.serialize(write)?;
```
Comment (PSeitz) [inline fastfield_codecs/src/multilinearinterpol_2.rs]: actually we could get the max vals and apply bitpacking to save some more bits (except from the f32)

(store the num bits in the codec header)

Comment (fmassot) [reply fastfield_codecs/src/multilinearinterpol_2.rs]: I won't do that for now :)

File: fastfield_codecs/src/multilinearinterpol_2.rs (hunk 4 above)
Comment (fmassot) [reply fastfield_codecs/src/multilinearinterpol_2.rs]: Yes indeed, I replaced it with `first_value`.

**Review by fulmicoton — APPROVED** [review]

File: fastfield_codecs/src/frame_of_reference.rs [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+use tantivy_bitpacker::BitPacker;
+
+use common::BinarySerializable;
+use common::DeserializeFrom;
+use tantivy_bitpacker::BitUnpacker;
+
+const BLOCK_SIZE: u64 = 128;
+
+#[derive(Clone)]
+pub struct FrameOfReferenceFastFieldReader {
```
Comment (fulmicoton) [inline fastfield_codecs/src/frame_of_reference.rs]: I think we can use the abbreviation `FOR` for once :)
It's one of those acronyms that is more famous than the underlying expression.

File: fastfield_codecs/src/frame_of_reference.rs [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+
+impl FastFieldCodecReader for FrameOfReferenceFastFieldReader {
+    /// Opens a fast field given a file.
+    fn open_from_bytes(bytes: &[u8]) -> io::Result<Self> {
+        let footer_len: u32 = (&bytes[bytes.len() - 4..]).deserialize()?;
+        let (_, mut footer) = bytes.split_at(bytes.len() - (4 + footer_len) as usize);
+        let footer = FrameOfReferenceFooter::deserialize(&mut footer)?;
+        let mut block_readers = Vec::with_capacity(footer.block_metadatas.len());
+        let mut current_data_offset = 0;
+        for block_metadata in footer.block_metadatas.into_iter() {
```
Comment (fulmicoton) [inline fastfield_codecs/src/frame_of_reference.rs]:
```suggestion
        for block_metadata in footer.block_metadatas {
```

File: fastfield_codecs/src/frame_of_reference.rs [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+        Ok(Self {
+            num_vals: footer.num_vals,
+            min_value: footer.min_value,
+            max_value: footer.max_value,
+            block_readers,
+        })
+    }
+
+    #[inline]
+    fn get_u64(&self, doc: u64, data: &[u8]) -> u64 {
```
Comment (fulmicoton) [inline fastfield_codecs/src/frame_of_reference.rs]: doc -> DocId (which is a u32 actually)

File: fastfield_codecs/src/multilinearinterpol_v2.rs [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+use tantivy_bitpacker::BitPacker;
+
+use common::BinarySerializable;
+use common::DeserializeFrom;
+use tantivy_bitpacker::BitUnpacker;
+
+const BLOCK_SIZE: u64 = 512;
+
+#[derive(Clone)]
+pub struct MultiLinearInterpolV2FastFieldReader {
```
Comment (fulmicoton) [inline fastfield_codecs/src/multilinearinterpol_v2.rs]: I suggest we stop calling this MultiLinear. MultiLinear usually means linear against with regard to several variables.
PiecewiseLinear is the correct terminology I believe.

File: fastfield_codecs/src/frame_of_reference.rs (hunk 8 above)
Comment (fmassot) [reply fastfield_codecs/src/frame_of_reference.rs]: This is defined at the trait level, I prefer to change that in another PR.

Comment (fulmicoton) [reply fastfield_codecs/src/frame_of_reference.rs]: Actually scratch that!

This is not necessarily a doc here, which is why this is u64.
The multivalued fast field rely on 2 fast field as you know.

In this case the second one is *not* a DocId. Let's call doc "idx" or something in this function.
We should fix other fastfields as well as the trait.

Comment (fmassot) [reply fastfield_codecs/src/frame_of_reference.rs]: Ok I totally forgot about multivalued ff, I updated the trait.

**Comment by fmassot** [comment]
@PSeitz @fulmicoton : here is my take on what to do for next tantivy release and what to do after:

These remarks are valid for integers only (for floats, our codecs are useless):
- piecewise linear interpolation should have the best compression ratio for monotonically increasing series
- bitpacking should have the best compression ratio on random stuff
- frame of reference (FOR) should have the best compression for series in between

What can we do for tantivy next release?
- unplug linear interpolation and replace multilinear interpolation by new piecewise linear interpolation (the bug fix increases a little bit the compression ratio but nothing noteworthy)
- update the benchmark with new datasets
- I don't want to plug immediately the FOR stuff because I think we can do slightly better and I would also want to activate piecewise interpolation only on monotonically increasing series, I would introduce that later.

Future work:
- finalize the FOR codecs, activate piecewise linear for only monotonically increasing series
- sparse codecs
- float codecs

**Comment by fulmicoton** [comment]
@fmassot agreeing with  the above.

**Comment by fulmicoton** [comment]
closing as obsolete. it was going in the right direction IMHO.

**Pull request closed without merging**